bootstrap/bpdoc/reader.go

```diff
@@ -0,0 +1,511 @@
+// Copyright 2019 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpdoc
+
+import (
+	"fmt"
+	"go/ast"
+	"go/doc"
+	"go/parser"
+	"go/token"
+	"html/template"
+	"reflect"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/google/blueprint/proptools"
+)
+
+type Reader struct {
+	pkgFiles map[string][]string // Map of package name to source files, provided by constructor
+
+	mutex sync.Mutex
+	pkgs  map[string]*doc.Package    // Map of package name to parsed Go AST, protected by mutex
+	ps    map[string]*PropertyStruct // Map of type name to property struct, protected by mutex
+}
+
+func NewReader(pkgFiles map[string][]string) *Reader {
+	return &Reader{
+		pkgFiles: pkgFiles,
+		pkgs:     make(map[string]*doc.Package),
+		ps:       make(map[string]*PropertyStruct),
+	}
+}
+
+func funcToPkgPath(f string) (string, error) {
+	p := strings.Split(f, "/")
+
+	pkgPath := strings.Join(p[:len(p)-1], "/")
+
+	p = strings.Split(p[len(p)-1], ".")
+	if len(p) < 2 {
+		return "", fmt.Errorf("failed to extract package path from %q", f)
+	}
+
+	if pkgPath != "" {
+		pkgPath += "/"
+	}
+	pkgPath += p[0]
+
+	return pkgPath, nil
+}
+
+func (r *Reader) ModuleType(name string, factory reflect.Value) (*ModuleType, error) {
+	f := runtime.FuncForPC(factory.Pointer())
+
+	pkgPath, err := funcToPkgPath(f.Name())
+	if err != nil {
+		return nil, err
+	}
+
+	factoryName := strings.TrimPrefix(f.Name(), pkgPath+".")
+
+	text, err := r.moduleTypeDocs(pkgPath, factoryName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ModuleType{
+		Name:    name,
+		PkgPath: pkgPath,
+		Text:    text,
+	}, nil
+}
+
+func (r *Reader) Package(pkgPath string) (*Package, error) {
```

Comment:
In a similar vein, having this function here in reader.go when the definition of the Package struct itself and all the callers of this function are in bpdoc feels unnatural to me. The way I see it, reader.go is a sort of lower-level layer that handles more primitive actions, and making it handle higher-level structures blurs the border.